docs: adds documentation for personal api tokens

sighphyre opened the pull request:
This adds documentation for what the new personal api tokens are, and where to use them. As well as a how to guide on creating them

sighphyre commented on website/docs/user_guide/token.mdx:
```diff
@@ -5,7 +5,7 @@ title: How to create API Tokens
 import newTokenImage from '/img/admin_create_token_form.png';
 
-::: info Permissions
+:::info Permissions
```
Not really relevant to this PR but it was broken and seemed overkill to make a new branch. Seems to be an autoformatter gone wild. Currently not working on the deployed version either: https://docs.getunleash.io/user_guide/api-token

thomasheartman replied:
Great, thanks! 😄

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
```diff
@@ -18,9 +18,10 @@ This section describes what API tokens are. For information on how to create the
 :::
 
-Use API tokens to connect to the Unleash server API. API tokens come in three distinct types:
+Use API tokens to connect to the Unleash server API. API tokens come in fours distinct types:
```
```suggestion
Use API tokens to connect to the Unleash server API. API tokens come in four distinct types:
```

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
```diff
@@ -18,9 +18,10 @@ This section describes what API tokens are. For information on how to create the
-Use API tokens to connect to the Unleash server API. API tokens come in three distinct types:
+Use API tokens to connect to the Unleash server API. API tokens come in fours distinct types:
 
 -   [Admin tokens](#admin-tokens)
+-   [Personal access tokens](#personal-access-tokens)
 -   [Client tokens](#client-tokens)
 -   [Front-end tokens](#front-end-tokens)
```
I think this list should be sorted alphabetically, but am happy to be challenged on that.


```suggestion
-   [Admin tokens](#admin-tokens)
-   [Client tokens](#client-tokens)
-   [Front-end tokens](#front-end-tokens)
-   [Personal access tokens](#personal-access-tokens)
```

sighphyre replied:
So the rationale here was that pats and admin tokens are so similar I wanted them grouped - happy to sort this alphabetically if you think that's clearer?

thomasheartman replied:
That was sorta what I figured. I think the list is short enough that either can work, so I'm not super opposed to leaving it as it was. Another option is putting the list in the same order the sections appear in the text. That might be the most sensible?

sighphyre replied:
I'm happy if you're happy. It's already in that order so I'm gonna close this comment thread if you're okay

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
```diff
@@ -52,6 +53,20 @@ Do **not** use admin tokens for:
+
+Use personal access tokens to:
+
+- Provide more fine grained permissions for automation than an admin token provides
```
```suggestion
- Provide more fine-grained permissions for automation than an admin token provides
```

thomasheartman commented on website/docs/reference/api/unleash/sidebar.js:
```diff
@@ -0,0 +1 @@
+module.exports = [{"type":"category","label":"Addons","link":{"type":"doc","id":"reference/api/unleash/addons"},"items":[{"type":"doc","id":"reference/api/unleash/get-addons","label":"getAddons","className":"api-method get"},{"type":"doc","id":"reference/api/unleash/create-addon","label":"createAddon","className":"api-method post"},{"type":"doc","id":"reference/api/unleash/get-addon","label":"get ...
```
Sorry, I think this is due to a misformed .gitconfig file (which has been fixed on main now). This is a generated file and should _not_ be added to git.

sighphyre replied:
Ah drat, thanks for the catch, I've removed this

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
````diff
@@ -141,6 +158,10 @@ Some example client tokens are:
     ```
     *:development.be44368985f7fb3237c584ef86f3d6bdada42ddbd63a019d26955178
     ```
+-   A personal access token with access to everything the user that created them has:
````
I don't think we need to repeat how PATs work here. It's mentioned for the other tokens because their form change depending on the details, but if I understand correctly, PATs are always the same: `user:<hash>`, yeah?

```suggestion
-   A personal access token:
```

sighphyre replied:
Yeah, they're always the same thing, sounds good, committed!

thomasheartman commented on website/docs/user_guide/personal-api-token.mdx:
```diff
@@ -0,0 +1,42 @@
+---
+id: personal-api-token
```
I suggest removing ids

```suggestion
```

thomasheartman commented on website/docs/user_guide/personal-api-token.mdx:
```diff
@@ -0,0 +1,42 @@
+
+:::
+
+Personal API tokens are a tool to enable a user to use the Admin API as themselves with their own set of permissions, rather than using an admin token. See [_how to use the Admin API_](../advanced/api_access) for more information.
```
I suggest linking to the reference docs. I'm also hesitant about linking to the api-access doc. I think it's a bit outdated. Do you think this would be sufficient?

```suggestion
[Personal API tokens](../reference/api-tokens-and-client-keys.mdx#personal-access-tokens) enable a user to use the Unleash admin API as themselves instead of as an global "admin" user.
```

sighphyre replied:
Yeah, I think that looks good, this is probably me getting too excited about adding links

thomasheartman commented on website/docs/user_guide/personal-api-token.mdx:
```diff
@@ -0,0 +1,42 @@
+
+![The user profile page with the "Personal API Tokens" menu item highlighted.](/img/create-pat-2.png)
+
+## Step 2: Navigate to New Token {#step-2}
+
+Navigate to "New Token".
+
+![The New Token element highlighted. ](/img/create-pat-3.png)
+
+## Step 3: Fill in the create personal API token form and create it {#step-3}
+
+Give your token a description and optionally set an expiry date. By default the expiry date is set to 30 days.
+
+![The New Token form with the description text box and create element highlighted. ](/img/create-pat-4.png)
+
+## Step 4: Save your token {#step-4}
```
I think these two steps can be joined together.

```suggestion
## Step 2: Create a personal API token {#step-2}

Access the token creation form using the "New token" button.

![The New Token element highlighted. ](/img/create-pat-3.png)

Give your token a description and optionally set an expiry date. The default expiry date is set to 30 days from creation.

![The New Token form with the description text box and create element highlighted. ](/img/create-pat-4.png)

## Step 3: Save your token {#step-3}
```

thomasheartman commented on website/docs/user_guide/personal-api-token.mdx:
```diff
@@ -0,0 +1,42 @@
+---
```
Can you move and rename this file to: `website/docs/how-to/how-to-create-personal-api-tokens.mdx`, please?

It's a how-to guide, so that's where it should go 🤓

sighphyre replied:
Definitely willing and able. Honestly this feels like good practise at getting more familiar with the docs which is something I want to do anyway

...but given how much it requires I'm also feeling separate PR

thomasheartman commented on website/docs/advanced/api_access.md:
```diff
@@ -7,7 +7,7 @@ It is possible to integrate directly with the Admin API. In this guide we will e
 ## Step 1: Create API token {#step-1-create-api-token}
 
-Please refer to [_how to create API tokens_](../user_guide/api-token) on how to create an API token. You'll need a token with `Admin` level access for this to work.
+Please refer to [_how to create API tokens_](../user_guide/api-token) on how to create an API token. Alternatively, you can use a [personal access token](/reference/api-tokens-and-client-keys#personal-access-tokens), see [_creating a personal api token_](../user_guide/personal-api-token) to create your own tokens. You'll need either an `Admin` or `Personal API Token` for this to work.
```
Links should also include file extensions

```suggestion
Please refer to [_how to create API tokens_](../user_guide/api-token) on how to create an API token. Alternatively, you can use a [personal access token](/reference/api-tokens-and-client-keys.mdx#personal-access-tokens), see [_creating a personal api token_](../user_guide/personal-api-token.mdx) to create your own tokens. You'll need either an `Admin` or `Personal API Token` for this to work.
```

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
```diff
@@ -32,10 +33,10 @@ An API token contains the following pieces of information:
-| Type | What kind of token it is: admin, client, or front-end. |
-| Projects | What projects a token has access to. |
-| Environment | What environment the token has access to. |
+| Token name (sometimes called "username") | The token's name. Names are **not** required to be unique. |
+| Type | What kind of token it is: admin, client, user (personal api token) or front-end. |
+| Projects | What projects a token has access to - not present in personal api tokens. |
+| Environment | What environment the token has access to - not present in personal api tokens. |
```
I wonder if personal api tokens should have a different table: they don't have projects or environments, but they do have an additional expiration date that the others do not. Thoughts?

sighphyre replied:
Ye, that's a pretty good idea actually, let me see what I can do here

sighphyre replied:
...actually on second thought, I'm not sure that PATs want their own table - they don't actually include any of the information described in the table. Maybe just a note explaining their format? WDYT?

thomasheartman replied:
Interesting. Do personal API tokens have names? I know that GitHub lets you name them (so you can have different ones for different purposes). But I think you're going in the right direction. In that case, we should probably point out that the table only applies to the other API tokens and ask the user to refer to the PAT section for info on them.

sighphyre replied:
They have a description but not a name, serves the same purpose as GH I think.

I've just added some text, but I haven't linked it to anything - we could link to the section higher up but that feels redundant, or the how to guide but I'm not sure that makes life better. Will you give it some eyeballs and see how you feel?

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
```diff
@@ -52,6 +53,20 @@ Do **not** use admin tokens for:
 
+### Personal access tokens
+
+**Personal access tokens** are a special form of admin tokens and grant access to the same resources that the user that created them has access to. These permissions are dynamic, so if a user's permissions change through addition of a custom role, the token will likewise have altered permissions. Using a personal access token to modify resources will reflect the creating user's name in the event  ...
```
Some minor rewording and breakup suggestions

```suggestion
**Personal access tokens** are a special form of admin tokens and grant access to the same resources that the user that created them has access to. These permissions are dynamic, so if a user's permissions change through addition of a custom role, the token will likewise have altered permissions. 

When using a personal access token to modify resources, the event log will list the token creator's name for that operation. 

All personal access tokens have a lifetime and **will stop working after the expiration** date.
```

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
```diff
@@ -127,6 +142,8 @@ The **environment** is the name of an environment on your Unleash server, such a
 
 The **hash** is 64 character long hexadecimal string.
 
+Personal API tokens do not have either the project or environment parts, since they mimic the user that created them. In place of the project field, the string `user` is present so that the token can be identified as a personal access token.
```
```suggestion
Personal API tokens do not contain project or environment information, since they mimic the user that created them. Instead, the token starts with the string `user`.
```

thomasheartman commented on website/sidebars.js:
```diff
@@ -50,6 +50,7 @@ module.exports = {
                     label: 'Unleash API guides',
                     items: [
                         'user_guide/api-token',
+                        'user_guide/personal-api-token',
```
```suggestion
                        'how-to/how-to-create-personal-api-tokens'
```

thomasheartman reviewed (changes requested):
Very nice work! 💯 I've got some minor comments, but they're almost all about formatting, file naming, and other nits that I don't expect anyone to be aware of (I should really get around to writing that "how to document" guide".

Have a look at the suggested changes, let me know what you think, and we'll take it from there 😄

thomasheartman commented on website/docs/advanced/api_access.md:
```diff
@@ -7,7 +7,7 @@ It is possible to integrate directly with the Admin API. In this guide we will e
 ## Step 1: Create API token {#step-1-create-api-token}
 
-Please refer to [_how to create API tokens_](../user_guide/api-token) on how to create an API token. You'll need a token with `Admin` level access for this to work.
+Please refer to [_how to create API tokens_](../user_guide/api-token) on how to create an API token. Alternatively, you can use a [personal access token](/reference/api-tokens-and-client-keys.mdx#personal-access-tokens), see [_creating a personal api token_](../user_guide/personal-api-token.mdx) to create your own tokens. You'll need either an `Admin` or `Personal API Token` for this to work.
```
I wonder if this can be rewritten a little bit. Something like:

```suggestion
You'll need either an [admin token](../reference/api-tokens-and-client-keys.mdx#admin-tokens) or a [personal API token](../reference/api-tokens-and-client-keys.mdx#personal-access-tokens) for this to work. To create one, follow the steps in the [_how to create API tokens_](../user_guide/token.mdx) guide or the [_how to create personal access tokens_](../user_guide/personal-api-token.mdx) guide, respectively.
```

This also links directly to the file for the token guide, which is preferable.

sighphyre replied:
I like it!

thomasheartman reviewed (commented):
Added a little extra comment on a thing that I think can be rephrased a bit. Let me know what you think.

Also, there are some inconsistencies regarding what we refer to the PATs as. Sometimes it's "personal access tokens", other times it's "personal API tokens". I think we should align on one or the other. Got a preference?

sighphyre commented:
> Added a little extra comment on a thing that I think can be rephrased a bit. Let me know what you think.
> 
> Also, there are some inconsistencies regarding what we refer to the PATs as. Sometimes it's "personal access tokens", other times it's "personal API tokens". I think we should align on one or the other. Got a preference?

Ye, like it, I've added it in.

Also good call. I've changed everything to `personal access tokens`, which is what GH calls them. I think that's correct anyways. How do you feel?

thomasheartman commented on website/docs/reference/api-tokens-and-client-keys.mdx:
```diff
@@ -18,25 +18,28 @@ This section describes what API tokens are. For information on how to create the
 | Projects | What projects a token has access to. |
 | Environment | What environment the token has access to. |
 
+Personal access tokens follow their own special format, and only contain an optional description for the token and an expiry date.
```
Thanks for this! There's a few things that are still a little unclear to me. I've also changed the description to a table for consistency. What do you think? And is it correct?


Also, when you create a PAT via the API, what endpoint is it? What sort of payload do you use?

```suggestion
Personal access tokens follow a different format and consist of the following pieces of information: 

| Name | Description |
| --- | --- |
| Description | A short description of what the token is to be used for. Must be unique. |
| Expiry data | The time when the token will stop working. |
```

sighphyre replied:
It doesn't actually need to be a unique description right now but I think that's a bug actually. I think your description of how this works what it should be - a unique description. I'm okay with taking your text as is and patching PATs to take on this behaviour before we release it

You can create a PAT via the API. I'm not really sure you should though, because it'd be a pain. You'd need to do one of two things, either get a a cookie via the login mechanism that the front end uses and use that in your request or get a PAT and use that PAT in your request to create a new PAT.

If you can get one of those two tokens, you can call {{base_url}}/api/admin/user/tokens with the following packet:

```
{
    "description": "Some stuff",
    "expiresAt": "2022-12-09T09:41:21.705Z"
}
```

Is this something we want to document and encourage?

thomasheartman replied:
>  I'm okay with taking your text as is and patching PATs to take on this behaviour before we release it

Yeah, I'm down with that! I think making it unique is sensible if that is the only human-readable identifier we have. In that case, it's kind of like a "name"?

> I'm not really sure you should though, because it'd be a pain. 

Right, because a PAT can only be created for your own user. An admin can't create a PAT **for** someone, right?

The question was mostly to ensure that it's a separate endpoint, so the `type`, and `projects` etc. doesn't really come into play at all.

sighphyre replied:
Nah, you gotta be you to make a token for yourself. The admin user can make a token for the admin user but you can't make one on behalf of someone else (at least not not and I don't think we have plans to do so). It's very much a separate endpoint, for the same reasons you point out - it would be chaos to use the existing ones

thomasheartman commented on website/docs/user_guide/personal-access-token.mdx:
```diff
@@ -0,0 +1,39 @@
+---
+title: How to create Personal Access Tokens
+---
+
```
Is this a duplicate of how-to-create-personal-access-tokens?

sighphyre replied:
Yeah, removed

thomasheartman commented on website/docs/reference/api/unleash/sidebar.js:
```diff
@@ -0,0 +1 @@
+module.exports = [{"type":"category","label":"Addons","link":{"type":"doc","id":"reference/api/unleash/addons"},"items":[{"type":"doc","id":"reference/api/unleash/get-addons","label":"getAddons","className":"api-method get"},{"type":"doc","id":"reference/api/unleash/create-addon","label":"createAddon","className":"api-method post"},{"type":"doc","id":"reference/api/unleash/get-addon","label":"get ...
```
This file should not be included in git. Could you remove it?

sighphyre replied:
Removed it again, also added it to the .gitignore this time, turns out its not actually in main

thomasheartman replied:
Huh, thought that was fixed in https://github.com/Unleash/unleash/pull/2070/files 🤔

sighphyre replied:
That nested .gitignore seems to be completely ignored by my git for some reason. Not sure why but adding it to the base .gitignore seems to sort it out

sighphyre commented on website/docs/advanced/api_access.md:
```diff
@@ -7,7 +7,7 @@ It is possible to integrate directly with the Admin API. In this guide we will e
 ## Step 1: Create API token {#step-1-create-api-token}
 
-Please refer to [_how to create API tokens_](../user_guide/api-token) on how to create an API token. You'll need a token with `Admin` level access for this to work.
+You'll need either an [admin token](../reference/api-tokens-and-client-keys.mdx#admin-tokens) or a [personal API token](../reference/api-tokens-and-client-keys.mdx#personal-access-tokens) for this to work. To create one, follow the steps in the [_how to create API tokens_](../user_guide/token.mdx) guide or the [_how to create personal access tokens_](../user_guide/personal-access-token.mdx) guide ...
```
Damnit I missed one, thanks!

thomasheartman commented on .gitignore:
```diff
@@ -3,6 +3,7 @@ lerna-debug
 npm-debug
 .DS_Store
 /dist
+website/docs/reference/api/unleash/sidebar.js
```
Oh, I get it now. It was in the docs .gitignore and _not_ the main .gitignore. Because .gitignores don't stack, it needs to be here too?

Suggest changing it to this for consistency, though. Might be more sidebars there later.

```suggestion
website/docs/reference/api/**/sidebar.js
```

thomasheartman replied:
Or wait ... **do** gitignores stack?

sighphyre replied:
Sure good call

sighphyre replied:
I think that depends on how you have them configured. It might be a my git being a git issue but I think if it keeps coming back on my machine it's worth adding it to the base config anyway, it's going to drive someone else mad too

sighphyre replied:
Could also be the way I'm adding my files, but same issue - someone's going to get hit by this

thomasheartman reviewed (approved):
Nice one! 🙌🏼  I think I'm good now, yeah 😄
